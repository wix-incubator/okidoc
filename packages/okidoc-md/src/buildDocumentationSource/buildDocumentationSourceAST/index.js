import buildDocumentationSourceAST from './buildDocumentationSourceAST';

export default buildDocumentationSourceAST;
