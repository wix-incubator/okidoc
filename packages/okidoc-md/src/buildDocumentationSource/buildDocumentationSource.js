import generate from '@babel/generator';

import buildDocumentationSourceAST from './buildDocumentationSourceAST';

function buildDocumentationSource({ entry, source, pattern, tag, visitor }) {
  return generate(
    buildDocumentationSourceAST({
      entry,
      source,
      pattern,
      tag,
      visitor,
    }),
  ).code;
}

export default buildDocumentationSource;
