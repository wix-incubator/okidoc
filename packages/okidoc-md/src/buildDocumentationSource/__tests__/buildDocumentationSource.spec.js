import path from 'path';
import buildDocumentationSource from '../';

describe('buildDocumentationSource', () => {
  describe('for API class', () => {
    describe('with @doc tag on class members', () => {
      it('should extract methods with tag', () => {
        const sourceCode = `
      class MyComponent {
        /**
         * show
         *
         * @doc UI
         * @example
         * component.show()
         */
        show() {
          console.log('show');
        }
      
        /**
         * Disable
         */
        disable() {
            console.log('disable');
        }
      }`;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            tag: 'UI',
          }),
        ).toMatchSnapshot();
      });

      it('should extract getters and setters with tag', () => {
        const sourceCode = `
      class MyComponent {
        /**
        * node getter
        */
        get node() {}

        /**
         * \`width\` getter
         *
         * @doc UI
         */
        get width() {
          console.log('get width');
        }
        
        /**
         * \`width\` setter
         *
         * @doc UI
         */
        set width(value) {
          console.log('set width', value);
        }
      }`;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            tag: 'UI',
          }),
        ).toMatchSnapshot();
      });

      it('should ignore private methods with tag', () => {
        const sourceCode = `
      class MyComponent {
        /**
         * init
         *
         * @doc UI
         */
        private init() {
          console.log('init');
        }
      }`;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            tag: 'UI',
          }),
        ).toMatchSnapshot();
      });

      it('should ignore methods with different tag', () => {
        const sourceCode = `
      class MyComponent {
        /**
         * show
         *
         * @doc UI
         */
        show() {
          console.log('show');
        }
      
        /**
         * disable
         *
         * @doc Destroy
         */
        destroy() {
            console.log('destroy');
        }
      }`;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            tag: 'UI',
          }),
        ).toMatchSnapshot();
      });
    });

    describe('with @doc tag on class declaration', () => {
      it('should extract class', () => {
        const sourceCode = `
      /**
      * @doc UI
      */
      class MyComponent extends View {
        view: View;

        /**
        * Creates a component
        */
        constructor(config: any) {}

        /**
        * \`isHidden\` getter
        */
        get isHidden() {}

        /**
         * show
         */
        show() {
          console.log('show');
        }
      
        /**
         * hide
         */
        hide() {
          console.log('hide');
        }
        
        disable() {
          console.log('hide');
        }
      }`;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            tag: 'UI',
          }),
        ).toMatchSnapshot();
      });

      it('should ignore class private members', () => {
        const sourceCode = `
      /**
      * @doc UI
      */
      class MyComponent {      
        /**
        * init
        */
        private init() {}
      
        render() {
          console.log('render');
        }
      }`;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            tag: 'UI',
          }),
        ).toMatchSnapshot();
      });
    });

    describe('with custom visitor', () => {
      it('should extract class members', () => {
        const sourceCode = `
        function api() {}
        
        class MyComponent {
          @api()
          get isHidden() {}
  
          /**
           * show
           */
          @api()
          show() {
            console.log('show');
          }
        
          /**
           * hide
           */
          @api()
          hide() {
            console.log('hide');
          }
          
          disable() {
            console.log('hide');
          }
        }`;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            visitor: require.resolve('./fixtures/customApiVisitor.js'),
          }),
        ).toMatchSnapshot();
      });
    });

    describe('should cleanup typescript vs flow difference', () => {
      it('interface void method', () => {
        const sourceCode = `
        interface IView {
          show();
        }
      `;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            tag: 'UI',
          }),
        ).toMatchSnapshot();
      });

      it('interface with optional method', () => {
        const sourceCode = `
        interface IView {
          show?();
        }
      `;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            tag: 'UI',
          }),
        ).toMatchSnapshot();
      });

      it('class public method', () => {
        const sourceCode = `
        /**
        * @doc UI
        */
        class View {
          public show() {}
        }
      `;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            tag: 'UI',
          }),
        ).toMatchSnapshot();
      });

      it('class constructor parameter property', () => {
        const sourceCode = `
        /**
        * @doc UI
        */
        class View {
          constructor(private config) {}
        }
      `;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            tag: 'UI',
          }),
        ).toMatchSnapshot();
      });
    });
  });

  describe('for functions', () => {
    describe('with @doc tag on function declaration', () => {
      it('should extract functions with tag', () => {
        const sourceCode = `
      /**
       * initComponent
       */
      function initComponent() {
        console.log('initComponent');
      }

      /**
       * showComponent
       *
       * @doc UI
       */
      function showComponent() {
        console.log('showComponent');
      }
      
      /**
       * hideComponent
       *
       * @doc UI
       */
      function hideComponent() {
        console.log('hideComponent');
      }
      `;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            tag: 'UI',
          }),
        ).toMatchSnapshot();
      });
    });

    describe('with @doc tag on var/let/const declaration', () => {
      it('should extract functions with tag', () => {
        const sourceCode = `
      /**
       * funcToIgnore
       */
      var funcToIgnore = function () {
        console.log('funcToIgnore');
      }

      /**
       * varFunc
       * @doc UI
       */
      var varFunctionExpression = function () {
        console.log('varFunc');
      }
      
      /**
       * varArrowFunc
       * @doc UI
       */
      var varArrowFunctionExpression = () => {
        console.log('varArrowFunc');
      }

      /**
       * letArrowFunc
       * @doc UI
       */
      let letArrowFunctionExpression = () => {
        console.log('letArrowFunc');
      }
      
      /**
       * constArrowFunc
       * @doc UI
       */
      const constArrowFunctionExpression = () => {
        console.log('constArrowFunc');
      }
      `;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            tag: 'UI',
          }),
        ).toMatchSnapshot();
      });
    });

    describe('with @doc tag on export declaration', () => {
      it('should extract functions with tag', () => {
        const sourceCode = `
      /**
       * f1 description
       * @doc UI
       */
      export function exportFunctionDeclaration(x) {}

      /**
       * f2 description
       * @doc UI
       */
      export const exportConstArrowFunctionExpression = (x) => {}
      
      /**
       * f3 description
       * @doc UI
       */
      export default exportDefaultConstArrowFunctionExpression = (x) => {}
      `;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            tag: 'UI',
          }),
        ).toMatchSnapshot();
      });
    });
  });

  describe('by source entry', () => {
    it('should extract documentation source from entry dependencies', () => {
      expect(
        buildDocumentationSource({
          entry: require.resolve('./fixtures/src/index.ts'),
          tag: 'UI',
        }),
      ).toMatchSnapshot();

      expect(
        buildDocumentationSource({
          entry: require.resolve('./fixtures/src/index.ts'),
          tag: 'Events',
        }),
      ).toMatchSnapshot();
    });

    it('should show readable error if source from entry is invalid', () => {
      expect(() => {
        buildDocumentationSource({
          entry: require.resolve('./fixtures/invalid-src/index.ts'),
          tag: 'Events',
        });
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('by source glob pattern', () => {
    it('should extract documentation source from files matched glob', () => {
      expect(
        buildDocumentationSource({
          pattern: `${path.join(__dirname, 'fixtures/src')}/**/*.ts`,
          tag: 'UI',
        }),
      ).toMatchSnapshot();

      expect(
        buildDocumentationSource({
          pattern: `${path.join(__dirname, 'fixtures/src')}/**/*.ts`,
          tag: 'utils',
        }),
      ).toMatchSnapshot();
    });

    it('should show readable error if source from matched file is invalid', () => {
      expect(() => {
        buildDocumentationSource({
          pattern: `${path.join(__dirname, 'fixtures/invalid-src')}/**/*.ts`,
          tag: 'Events',
        });
      }).toThrowErrorMatchingSnapshot();
    });
  });
});
