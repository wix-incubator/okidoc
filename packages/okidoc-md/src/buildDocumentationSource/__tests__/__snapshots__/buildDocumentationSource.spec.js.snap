// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildDocumentationSource by source entry should extract documentation source from entry dependencies 1`] = `
"/**
* View
*/
class View {
  /**
  * getter for \`isHidden\` value
  *
  * @example
  * console.log(component.isHidden) // false
  *
  * @returns \`true\` if hidden
  */
  get isHidden(): boolean {}

  /**
  * show
  *
  * @example
  * component.show()
  */
  show() {}

  /**
  * Hide
  *
  * @example
  * component.hide()
  */
  hide() {}

}

/**
* Component
*/
class Component {
  /** */
  constructor() {}

  /**
  * Get component node
  */
  getNode(): Node {}

  /**
  * Render component
  */
  render() {}

}"
`;

exports[`buildDocumentationSource by source entry should extract documentation source from entry dependencies 2`] = `
"/**
* Events
*/
class Events {
  /**
  * add event listener
  */
  addEventListener(fn: Function) {}

  /**
  * remove event listener
  */
  removeEventListener(fn: Function) {}

}"
`;

exports[`buildDocumentationSource by source entry should show readable error if source from entry is invalid 1`] = `
"packages/okidoc-md/src/buildDocumentationSource/__tests__/fixtures/invalid-src/index.ts: Unterminated string constant (1:22)
> 1 | import Component from './Component;
    |                      ^
  2 | 
  3 | export { Component };"
`;

exports[`buildDocumentationSource by source glob pattern should extract documentation source from files matched glob 1`] = `
"/**
* Component
*/
class Component {
  /** */
  constructor() {}

  /**
  * Get component node
  */
  getNode(): Node {}

  /**
  * Render component
  */
  render() {}

}

/**
* View
*/
class View {
  /**
  * getter for \`isHidden\` value
  *
  * @example
  * console.log(component.isHidden) // false
  *
  * @returns \`true\` if hidden
  */
  get isHidden(): boolean {}

  /**
  * show
  *
  * @example
  * component.show()
  */
  show() {}

  /**
  * Hide
  *
  * @example
  * component.hide()
  */
  hide() {}

}"
`;

exports[`buildDocumentationSource by source glob pattern should extract documentation source from files matched glob 2`] = `
"/**
* mountComponent
*/
function mountComponent(component: Function, node: Node) {}

/**
* renderComponent
*/
function renderComponent(component: Function) {}

/**
* unmountComponent
*/
function unmountComponent(component: Function) {}"
`;

exports[`buildDocumentationSource by source glob pattern should show readable error if source from matched file is invalid 1`] = `
"packages/okidoc-md/src/buildDocumentationSource/__tests__/fixtures/invalid-src/index.ts: Unterminated string constant (1:22)
> 1 | import Component from './Component;
    |                      ^
  2 | 
  3 | export { Component };"
`;

exports[`buildDocumentationSource for API class should cleanup typescript vs flow difference class constructor parameter property 1`] = `
"/** */
class View {
  /** */
  constructor(config) {}

}"
`;

exports[`buildDocumentationSource for API class should cleanup typescript vs flow difference class public method 1`] = `
"/** */
class View {
  /** */
  show() {}

}"
`;

exports[`buildDocumentationSource for API class should cleanup typescript vs flow difference interface void method 1`] = `
"/** */
interface IView {
  show(): void;
}"
`;

exports[`buildDocumentationSource for API class with @doc tag on class declaration should extract class 1`] = `
"/** */
class MyComponent {
  /** */
  view: View;

  /**
  * Creates a component
  */
  constructor(config: any) {}

  /**
  * \`isHidden\` getter
  */
  get isHidden() {}

  /**
  * show
  */
  show() {}

  /**
  * hide
  */
  hide() {}

  /** */
  disable() {}

}"
`;

exports[`buildDocumentationSource for API class with @doc tag on class declaration should ignore class private members 1`] = `
"/** */
class MyComponent {
  /** */
  render() {}

}"
`;

exports[`buildDocumentationSource for API class with @doc tag on class members should extract getters and setters with tag 1`] = `
"/** */
class API {
  /**
  * \`width\` getter
  */
  get width() {}

  /**
  * \`width\` setter
  */
  set width(value) {}

}"
`;

exports[`buildDocumentationSource for API class with @doc tag on class members should extract methods with tag 1`] = `
"/** */
class API {
  /**
  * show
  *
  * @example
  * component.show()
  */
  show() {}

}"
`;

exports[`buildDocumentationSource for API class with @doc tag on class members should ignore methods with different tag 1`] = `
"/** */
class API {
  /**
  * show
  */
  show() {}

}"
`;

exports[`buildDocumentationSource for API class with @doc tag on class members should ignore private methods with tag 1`] = `""`;

exports[`buildDocumentationSource for API class with custom visitor should extract class members 1`] = `
"/** */
class API {
  /** */
  get isHidden() {}

  /**
  * show
  */
  show() {}

  /**
  * hide
  */
  hide() {}

}"
`;

exports[`buildDocumentationSource for functions with @doc tag on export declaration should extract functions with tag 1`] = `
"/**
* f1 description
*/
function exportFunctionDeclaration(x) {}

/**
* f2 description
*/
function exportConstArrowFunctionExpression(x) {}

/**
* f3 description
*/
function exportDefaultConstArrowFunctionExpression(x) {}"
`;

exports[`buildDocumentationSource for functions with @doc tag on function declaration should extract functions with tag 1`] = `
"/**
* showComponent
*/
function showComponent() {}

/**
* hideComponent
*/
function hideComponent() {}"
`;

exports[`buildDocumentationSource for functions with @doc tag on var/let/const declaration should extract functions with tag 1`] = `
"/**
* varFunc
*/
function varFunctionExpression() {}

/**
* varArrowFunc
*/
function varArrowFunctionExpression() {}

/**
* letArrowFunc
*/
function letArrowFunctionExpression() {}

/**
* constArrowFunc
*/
function constArrowFunctionExpression() {}"
`;
